src/topojson/topojson.ml

```diff
@@ -1 +1,182 @@
+            else false
+          in
+          loop 0
+
+      let to_json arr = J.array J.float arr
+    end
+
+    module Arcs = struct
+      type t = int array
+
+      let to_json arr = J.array J.int arr
+    end
+
+    module Point = struct
+      type t = Position.t
+
+      let typ = "Point"
+      let position = Fun.id
+      let v position = position
+      let parse_coords coords = J.to_array (decode_or_err J.to_float) coords
+      let base_of_json json = parse_by_type json parse_coords typ
+
+      let to_json ?bbox ?(foreign_members = []) position =
+        J.obj
+          ([
+             ("type", J.string typ); ("coordinates", Position.to_json position);
+           ]
+          @ bbox_to_json_or_empty bbox
+          @ foreign_members)
+    end
+
+    module MultiPoint = struct
+      type t = Point.t array
+
+      let typ = "MultiPoint"
+      let coordinates = Fun.id
+      let v positions = positions
+
+      let parse_coords coords =
+        try J.to_array (decode_or_err Point.parse_coords) coords
+        with Failure m -> Error (`Msg m)
+
+      let base_of_json json = parse_by_type json parse_coords typ
+
+      let to_json ?bbox ?(foreign_members = []) positions =
+        J.obj
+          ([
+             ("type", J.string typ);
+             ("coordinates", J.array Position.to_json positions);
+           ]
+          @ bbox_to_json_or_empty bbox
+          @ foreign_members)
+    end
+
+    module LineString = struct
+      type t = int array
+
+      let typ = "LineString"
+      let v arcs = arcs
+
+      let to_json ?arcs ?bbox ?(foreign_members = []) arcs =
+        J.obj
+          ([ ("type", J.string typ); ("arcs", J.array Arcs.to_json arcs) ]
+          (* @ arcs *)
+          @ bbox_to_json_or_empty bbox
+          @ foreign_members)
+    end
+
+    module MultiLineString = struct
+      type t = LineString.t array
+
+      let typ = "MultiLineString"
+      let v arcs = arcs
+
+      let to_json ?arcs ?bbox ?(foreign_members = []) arcs =
+        J.obj
+          ([ ("type", J.string typ); ("arcs", J.array (J.array J.int) arcs) ]
+          (* @ arcs *)
+          @ bbox_to_json_or_empty bbox
+          @ foreign_members)
+    end
+
+    type geometry =
+      | Point of Point.t
+      | MultiPoint of MultiPoint.t
+      | LineString of LineString.t
+      | MultiLineString of MultiLineString.t
+    (* | Polygon of Polygon.t
+       | MultiPolygon of MultiPolygon.t
+       | Collection of t list *)
+
+    and t = {
+      geometry : geometry;
+      arcs : int array;
+      id : G.json option;
+      properties : (string * G.json) list;
+      foreign_members : (string * G.json) list;
+    }
+  end
+
+  module Topology = struct
+    type t = { objects : (string * G.json) list; arcs : int array array }
+
+    let of_json json =
+      match (J.find json [ "objects" ], J.find json [ "arcs" ]) with
+      | Some objects, arcs ->
+          Result.map (fun v -> { objects; arcs }) (J.to_obj v)
```

Comment:
This isn't quite right, and doesn't currently compile:

Something like this might be better (if you define the appropriate `let+` and `let*` or feel free to use `Result.map` and `Result.bind` if you prefer).

```ocaml
 let of_json json =
      match (J.find json [ "objects" ], J.find json [ "arcs" ]) with
      | Some objects, Some arcs -> (
        let* objects = J.to_obj objects in
        let+ arcs = J.to_array (decode_or_err (J.to_array (decode_or_err J.to_int))) arcs in
        { objects; arcs }
      )
      | _, _ -> Error (`Msg "No objects and/or arcs field in Topology object!")
````

Reply:
This seems somewhat newer to me the ```let*``` & ```let+``` things will explore more on this...
Thanks for letting me know these Patrick:))